File: backend/schemas/learning.py
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field


class TaskLevel(str, Enum):
    beginner = "beginner"
    intermediate = "intermediate"
    advanced = "advanced"


class TaskType(str, Enum):
    reading = "reading"
    practice = "practice"
    review = "review"
    quiz = "quiz"


class QuizStatus(str, Enum):
    in_progress = "in_progress"
    completed = "completed"


class ReviewStatus(str, Enum):
    pending = "pending"
    completed = "completed"


class GoalCreate(BaseModel):
    title: str
    subject: str | None = None
    topic: str | None = None
    level: TaskLevel = TaskLevel.beginner
    daily_minutes: int = Field(default=60, ge=1)
    duration_weeks: int = Field(default=12, ge=1)
    available_days: list[str] = ["monday", "tuesday", "wednesday", "thursday", "friday"]
    notes: str | None = None


class TaskUpdate(BaseModel):
    task_id: str | None = None
    status: str
    actual_duration_min: int | None = Field(default=None, ge=0)


class LearningStats(BaseModel):
    tasks_completed: int
    tasks_total: int
    study_minutes: int
    quiz_avg_score: float
    reviews_completed: int
    new_concepts_learned: int
    streak_days: int
    mastery_overall: float


class MasteryByTopic(BaseModel):
    topic: str
    mastery: float


class StudyDay(BaseModel):
    date: str
    minutes: int
    tasks: int


class LearningReportResponse(BaseModel):
    goal_id: str
    period: str
    stats: LearningStats
    mastery_by_topic: list[MasteryByTopic] = []
    weak_points: list[str] = []
    study_calendar: list[StudyDay] = []


class FlashcardResponse(BaseModel):
    flashcard_id: str
    front: str
    back: str
    tags: list[str] = []
    concept: str | None = None
    mastery: float = Field(ge=0.0, le=1.0)
    reviews_count: int = 0
    next_review_at: datetime | None = None


class ReviewItemResponse(BaseModel):
    review_id: str
    concept: str
    mastery: float
    last_reviewed: datetime | None = None
    interval_days: int = 1
    status: str
    new_interval_days: int = 0
    next_review_at: datetime | None = None
    easiness: float = 2.5


class GoalResponse(BaseModel):
    goal_id: str
    title: str
    status: str
    plan: dict | None = None
    progress_pct: float = 0.0
    streak_days: int = 0
    created_at: datetime | None = None


class GoalDetail(BaseModel):
    goal_id: str
    title: str
    status: str
    progress_pct: float
    current_week: int
    current_milestone: dict | None = None
    streak_days: int
    total_study_minutes: int = 0


class QuizStartRequest(BaseModel):
    topic: str | None = None
    question_count: int = Field(default=3, ge=1, le=20)


class QuizStartResponse(BaseModel):
    quiz_id: str
    questions: list[dict]


class QuizAnswerRequest(BaseModel):
    question_id: str
    answer: str


class QuizAnswerResponse(BaseModel):
    question_id: str
    score: int = Field(ge=0, le=5)
    max_score: int = 5
    feedback: str
    missing_concepts: list[str] = []
    weak_points: list[dict] = []
    next_review_at: datetime | None = None


class ReviewFeedbackRequest(BaseModel):
    quality: int = Field(ge=0, le=5)
    response_time_ms: int | None = Field(default=None, ge=0)


class FlashcardBatchResponse(BaseModel):
    batch_id: str
    source: str
    flashcards: list[FlashcardResponse]
    generated_count: int
    created_at: datetime | None = None


class MindMapGenerateRequest(BaseModel):
    source_type: str = "goal_topics"
    scope: str = "current_week"
    include_mastery: bool = True


class MindMapResponse(BaseModel):
    mindmap_id: str
    title: str
    format: str = "mermaid"
    content: str
    node_count: int = 0
    generated_at: datetime | None = None

File: frontend/src/api/learning.ts
import { api } from './client';

export async function createGoal(data: {
  title: string; subject?: string; topic?: string; level?: string;
  daily_minutes?: number; duration_weeks?: number; available_days?: string[]; notes?: string;
}) {
  return api.post('/learning/goals', data);
}

export async function listGoals() {
  return api.get<{ goals: Array<{ goal_id: string; title: string; status: string; progress_pct: number; streak_days: number; created_at: string }> }>('/learning/goals');
}

export async function getGoal(goalId: string) {
  return api.get(`/learning/goals/${goalId}`);
}

export async function getTodayTasks(goalId: string) {
  return api.get<{ date: string; tasks: Array<{ task_id: string; title: string; type: string; duration_min: number; status: string; order: number }>; total_minutes: number }>(`/learning/goals/${goalId}/tasks/today`);
}

export async function updateTask(taskId: string, status: string, actualDurationMin?: number) {
  return api.patch(`/learning/tasks/${taskId}`, { status, actual_duration_min: actualDurationMin });
}

export async function startQuiz(goalId: string, topic?: string, questionCount = 3) {
  return api.post<{ quiz_id: string; questions: Array<{ question_id: string; type: string; content: string; expected_concepts: string[] }> }>(
    `/learning/goals/${goalId}/quiz/start`,
    { topic, question_count: questionCount }
  );
}

export async function submitAnswer(quizId: string, questionId: string, answer: string) {
  return api.post<{ question_id: string; score: number; max_score: number; feedback: string; missing_concepts: string[]; weak_points: Array<{ concept: string; mastery: number }>; next_review_at: string }>(
    `/learning/quiz/${quizId}/answer`,
    { question_id: questionId, answer }
  );
}

export async function getReviews(goalId: string) {
  return api.get<{ due_reviews: Array<{ review_id: string; concept: string; mastery: number; interval_days: number; status: string }>; total_due: number }>(`/learning/goals/${goalId}/reviews`);
}

export async function submitReviewFeedback(reviewId: string, quality: number) {
  return api.post<{ review_id: string; concept: string; new_interval_days: number; next_review_at: string; easiness: number }>(
    `/learning/reviews/${reviewId}/feedback`,
    { quality }
  );
}

export async function submitFlashcardFeedback(flashcardId: string, quality: number) {
  return api.post<{ review_id: string; concept: string; new_interval_days: number; next_review_at: string; easiness: number }>(
    `/learning/flashcards/${flashcardId}/feedback`,
    { quality }
  );
}

export async function generateFlashcards(goalId: string, file: File | null, text: string | null, count = 10) {
  const fd = new FormData();
  if (file) fd.append('file', file);
  if (text) fd.append('text', text);
  fd.append('flashcard_count', String(count));
  return api.uploadFormData(`/learning/goals/${goalId}/flashcards/generate`, fd);
}

export async function listFlashcards(goalId: string, status?: string, limit = 50) {
  const params = new URLSearchParams({ limit: String(limit) });
  if (status) params.append('status', status);
  return api.get(`/learning/goals/${goalId}/flashcards?${params}`);
}

export async function generateMindmap(goalId: string, sourceType = 'goal_topics', scope = 'current_week') {
  return api.post<{ mindmap_id: string; title: string; format: string; content: string; node_count: number; generated_at: string }>(
    `/learning/goals/${goalId}/mindmap/generate`,
    { source_type: sourceType, scope, include_mastery: true }
  );
}

export async function getReport(goalId: string, period = 'week') {
  return api.get(`/learning/goals/${goalId}/report?period=${period}`);
}
